File: tests/clients/cs-client-tests/ClientTests.cs
using Moq;
using Moq.Protected;
using System.Net;
using System.Text;
using System.Text.Json;
using Xunit;

namespace Nexus.Client.Tests
{
    public class ClientTests
    {
        public const string NexusConfigurationHeaderKey = "Nexus-Configuration";

        [Fact]
        public async Task CanAuthenticateAndRefreshAsync()
        {
            // Arrange
            var messageHandlerMock = new Mock<HttpMessageHandler>();

            var tokenPair = new TokenPair(
                AccessToken: "123",
                RefreshToken: "456"
            );

            // -> get catalogs (1st try)
            var tryCount = 0;

            var catalogsResponseMessage1 = new HttpResponseMessage()
            {
                StatusCode = HttpStatusCode.Unauthorized
            };

            catalogsResponseMessage1.Headers.Add("WWW-Authenticate", "Bearer The token expired at ...");

            messageHandlerMock
                .Protected()
                .Setup<Task<HttpResponseMessage>>(
                    "SendAsync",
                    ItExpr.Is<HttpRequestMessage>(x => x.RequestUri!.ToString().Contains("catalogs") && tryCount == 0),
                    ItExpr.IsAny<CancellationToken>())
                .Callback<HttpRequestMessage, CancellationToken>((requestMessage, cancellationToken) =>
                {
                    var actual = requestMessage.Headers.Authorization!;
                    Assert.Equal($"Bearer {tokenPair.AccessToken}", $"{actual.Scheme} {actual.Parameter}");
                    catalogsResponseMessage1.RequestMessage = requestMessage;
                    tryCount++;
                })
                .ReturnsAsync(catalogsResponseMessage1);

            // -> refresh token
            var newTokenPair = new TokenPair(
                AccessToken: "123",
                RefreshToken: "456"
            );

            var refreshTokenResponseJsonString = JsonSerializer.Serialize(newTokenPair);

            var refreshTokenResponseMessage = new HttpResponseMessage(HttpStatusCode.OK)
            {
                Content = new StringContent(
                    refreshTokenResponseJsonString,
                    Encoding.UTF8, 
                    "application/json")
            };

            messageHandlerMock
                .Protected()
                .Setup<Task<HttpResponseMessage>>(
                    "SendAsync",
                    ItExpr.Is<HttpRequestMessage>(x => x.RequestUri!.ToString().EndsWith("refresh-token")),
                    ItExpr.IsAny<CancellationToken>())
                .Callback<HttpRequestMessage, CancellationToken>((requestMessage, cancellationToken) =>
                {
                    var refreshTokenRequest = JsonSerializer.Deserialize<RefreshTokenRequest>(requestMessage.Content!.ReadAsStream());
                    Assert.Equal(tokenPair.RefreshToken, refreshTokenRequest!.RefreshToken);
                })
                .ReturnsAsync(refreshTokenResponseMessage);

            // -> get catalogs (2nd try)
            var catalogId = "my-catalog-id";
            var expectedCatalog = new ResourceCatalog(Id: catalogId, default, default);

            var catalogsResponseMessage2 = new HttpResponseMessage()
            {
                StatusCode = HttpStatusCode.OK,
                Content = new StringContent(JsonSerializer.Serialize(expectedCatalog), Encoding.UTF8, "application/json"),
            };

            messageHandlerMock
                .Protected()
                .Setup<Task<HttpResponseMessage>>(
                    "SendAsync",
                    ItExpr.Is<HttpRequestMessage>(x => x.RequestUri!.ToString().Contains("catalogs") && tryCount == 1),
                    ItExpr.IsAny<CancellationToken>())
                .Callback<HttpRequestMessage, CancellationToken>((requestMessage, cancellationToken) =>
                {
                    var actual = requestMessage.Headers.Authorization!;
                    Assert.Equal($"Bearer {newTokenPair.AccessToken}", $"{actual.Scheme} {actual.Parameter}");
                })
                .ReturnsAsync(catalogsResponseMessage2);

            // -> http client
            var httpClient = new HttpClient(messageHandlerMock.Object)
            {
                BaseAddress = new Uri("http://localhost")
            };

            // -> API client
            var client = new NexusClient(httpClient);

            // Act
            client.SignIn(tokenPair);
            var actualCatalog = await client.Catalogs.GetCatalogAsync(catalogId);

            // Assert
            Assert.Equal(
                JsonSerializer.Serialize(expectedCatalog),
                JsonSerializer.Serialize(actualCatalog));
        }

        [Fact]
        public async Task CanAddConfigurationAsync()
        {
            // Arrange
            var messageHandlerMock = new Mock<HttpMessageHandler>();
            var catalogId = "my-catalog-id";
            var expectedCatalog = new ResourceCatalog(Id: catalogId, default, default);

            var actualHeaders = new List<IEnumerable<string>?>();

            messageHandlerMock
                .Protected()
                .Setup<Task<HttpResponseMessage>>(
                    "SendAsync",
                    ItExpr.IsAny<HttpRequestMessage>(),
                    ItExpr.IsAny<CancellationToken>())
                .Callback<HttpRequestMessage, CancellationToken>((requestMessage, cancellationToken) =>
                {
                    requestMessage.Headers.TryGetValues(NexusConfigurationHeaderKey, out var headers);
                    actualHeaders.Add(headers);
                })
                .ReturnsAsync(() =>
                {
                    return new HttpResponseMessage()
                    {
                        StatusCode = HttpStatusCode.OK,
                        Content = new StringContent(JsonSerializer.Serialize(expectedCatalog), Encoding.UTF8, "application/json"),
                    };
                });

            // -> http client
            var httpClient = new HttpClient(messageHandlerMock.Object)
            {
                BaseAddress = new Uri("http://localhost")
            };

            // -> API client
            var client = new NexusClient(httpClient);

            // -> configuration
            var configuration = new Dictionary<string, string>
            {
                ["foo1"] = "bar1",
                ["foo2"] = "bar2"
            };

            // Act
            _ = await client.Catalogs.GetCatalogAsync(catalogId);

            using (var disposable = client.AttachConfiguration(configuration))
            {
                _ = await client.Catalogs.GetCatalogAsync(catalogId);
            }

            _ = await client.Catalogs.GetCatalogAsync(catalogId);

            // Assert
            var encodedJson = Convert.ToBase64String(JsonSerializer.SerializeToUtf8Bytes(configuration));

            Assert.Collection(actualHeaders,
                headers => Assert.Null(headers),
                headers => Assert.Collection(headers, header => Assert.Equal(encodedJson, header)),
                headers => Assert.Null(headers));
        }
    }
}

File: tests/clients/python-client-tests/client-tests.py
import base64
import json

import pytest
from httpx import AsyncClient, MockTransport, Request, Response, codes
from nexusapi import NexusClient, ResourceCatalog, TokenPair

nexus_configuration_header_key = "Nexus-Configuration"

try_count1: int = 0

token_pair = TokenPair(
    access_token="123",
    refresh_token="456")

def _handler1(request: Request):
    global try_count1

    if "catalogs" in request.url.path:
        try_count1 += 1

        if try_count1 == 1:
            actual = request.headers["Authorization"]
            assert f"Bearer {token_pair.access_token}" == actual
            return Response(codes.UNAUTHORIZED, headers={"WWW-Authenticate" : "Bearer The token expired at ..."})

        else:
            catalog_json_string = '{"Id":"my-catalog-id","Properties":null,"Resources":null}'
            return Response(codes.OK, content=catalog_json_string)

    elif "refresh-token" in request.url.path:
        requestContent = request.content.decode("utf-8")
        assert '{"refreshToken": "456"}' == requestContent

        new_token_pair_json_string = '{ "accessToken": "123", "refreshToken": "456" }'
        return Response(codes.OK, content=new_token_pair_json_string)

    else:
        raise Exception("Unsupported path.")

@pytest.mark.asyncio
async def can_authenticate_and_refresh_test():

    # arrange
    catalog_id = "my-catalog-id"
    expected_catalog = ResourceCatalog(catalog_id, None, None)
    http_client = AsyncClient(base_url="http://localhost", transport=MockTransport(_handler1))

    async with NexusClient(http_client) as client:

        # act
        client.sign_in(token_pair)
        actual_catalog = await client.catalogs.get_async(catalog_id)
       
        # assert
        assert expected_catalog == actual_catalog

try_count2: int = 0

def _handler2(request: Request):
    global try_count2

    if "catalogs" in request.url.path:
        try_count2 += 1

        if (try_count2 == 1):
            assert not nexus_configuration_header_key in request.headers

        elif (try_count2 == 2):

            configuration = {
                "foo1": "bar1",
                "foo2": "bar2"
            }

            expected = base64.b64encode(json.dumps(configuration).encode("utf-8")).decode("utf-8")
            actual = request.headers[nexus_configuration_header_key]

            assert expected == actual

        elif (try_count2 == 3):
            assert not nexus_configuration_header_key in request.headers

        catalog_json_string = '{"Id":"my-catalog-id","Properties":null,"Resources":null}'
        return Response(codes.OK, content=catalog_json_string)

    elif "refresh-token" in request.url.path:
        requestContent = request.content.decode("utf-8")
        assert '{"refreshToken": "456"}' == requestContent

        new_token_pair_json_string = '{ "accessToken": "123", "refreshToken": "456" }'
        return Response(codes.OK, content=new_token_pair_json_string)

    else:
        raise Exception("Unsupported path.")

@pytest.mark.asyncio
async def can_add_configuration_test():

    # arrange
    catalog_id = "my-catalog-id"

    configuration = {
        "foo1": "bar1",
        "foo2": "bar2"
    }

    http_client = AsyncClient(base_url="http://localhost", transport=MockTransport(_handler2))

    async with NexusClient(http_client) as client:

        # act
        _ = await client.catalogs.get_async(catalog_id)

        with client.attach_configuration(configuration):
            _ = await client.catalogs.get_async(catalog_id)

        _ = await client.catalogs.get_async(catalog_id)

        # assert (already asserted in _handler2)
